### PR TITLE
Internal Client refactoring

### PR DESCRIPTION
**Pull request by yuce**
## Overview

Refactored all client code into InternalHTTPClient struct which implements InternalClient interface. This allowed us to get rid of some duplicate code and enables us to use another transport other than HTTP in the future.

Implements https://github.com/pilosa/pilosa/issues/894

## Pull request checklist

- [x] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [x] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [ ] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [x] I have resolved any merge conflicts.
- [ ] I have included tests that cover my changes.
- [x] All new and existing tests pass.

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -36,22 +36,22 @@ import (
...
+	options    *ClientOptions
 
 	// The client to use for HTTP communication.
 	HTTPClient *http.Client
 }
 
-// NewClient returns a new instance of Client to connect to host.
-func NewClient(host string, options *ClientOptions) (*Client, error) {
+// NewClient returns a new instance of InternalHTTPClient to connect to host.
+func NewClient(host string, options *ClientOptions) (*InternalHTTPClient, error) {
```
Comment (travisturner) [inline client.go]: Any reason not to call this `NewInternalHTTPClient()`?

File: client.go [hunk 2]
```diff
@@ -1229,3 +1205,27 @@ func nodePathToURL(node *Node, path string) url.URL {
 		Path:   path,
 	}
 }
+
+type InternalClient interface {
```
Comment (travisturner) [inline client.go]: I just want to leave a note here for discussion: typically an interface containing more than two or three methods is an indication that something hasn't been architected correctly. While I understand that putting the entire Client behind an interface might require this many methods, I don't want to let it go unquestioned.

File: executor.go [hunk 3]
```diff
@@ -47,26 +43,21 @@ type Executor struct {
...
+	// Client used for remote requests.
+	client InternalClient
 
 	// Maximum number of SetBit() or ClearBit() commands per request.
 	MaxWritesPerRequest int
 }
 
 // NewExecutor returns a new instance of Executor.
-func NewExecutor(clientOptions *ClientOptions) *Executor {
+func NewExecutor(clientOptions *ClientOptions) (*Executor, error) {
```
Comment (travisturner) [inline executor.go]: why did you add an `error` to the return here?

File: client.go (hunk 2 above)
Comment (yuce) [reply client.go]: Would it be OK if I added your comment to the interface and we would address that in another PR?

File: client.go (hunk 1 above)
Comment (yuce) [reply client.go]: I forgot to rename that function, going to do.

File: executor.go (hunk 3 above)
Comment (yuce) [reply executor.go]: I think I called a function which returned an `error` there and removed that code without without the `error`. Will remove it.

File: client.go [hunk 4]
```diff
@@ -61,13 +61,11 @@ func NewClient(host string, options *ClientOptions) (*Client, error) {
...
-	return NewClientFromURI(uri, options)
+	client := NewClientFromURI(uri, options)
+	return client, nil
 }
 
-func NewClientFromURI(uri *URI, options *ClientOptions) (*Client, error) {
-	if uri == nil {
-		return nil, ErrHostRequired
-	}
+func NewClientFromURI(defaultURI *URI, options *ClientOptions) *InternalHTTPClient {
```
Comment (travisturner) [inline client.go]: let's rename this one as well.

Comment (yuce) [reply client.go]: Oops. Updated.

**Comment by yuce** [comment]
@travisturner I want to implement https://github.com/pilosa/pilosa/issues/894 in order to remove workarounds for the client. Would you prefer that is implemented in this PR or another PR after this one is merged?

**Comment by travisturner** [comment]
@yuce either is fine with me. what's easier for you?

File: client.go (hunk 2 above)
Comment (travisturner) [reply client.go]: ok with me.

**Comment by yuce** [comment]
@travisturner Great, I'll push it to this branch.

**Comment by yuce** [comment]
Added scheme to the node status.
